### PR TITLE
[FAB-27] Waiting room

### PR DESCRIPTION
**Pull request by nicolauscg**
new features

- [x] on hub, add waiting room showing joined members
- [x] multipage story on hub
- [x] do needed changes for controller side 

out of scope: 
- timer stuff, plan to make those changes in another PR, also could make it easily switchable between finish button and timer, as not sure which one is better
- send canvas result through WS to server, will be in another PR too

### CODE REVIEW

**Comment by nicolauscg** [comment]
refactor of ws connection from ref to state is here https://github.com/deco-finter/fableous/pull/14/files/b865fb60948e9904c7d659e22d6225cadfe5d198..8a71368b23664a6c6d1d7b9a0051a62788abc780

**Comment by daystram** [comment]
> refactor of ws connection from ref to state is here https://github.com/deco-finter/fableous/pull/14/files/b865fb60948e9904c7d659e22d6225cadfe5d198..8a71368b23664a6c6d1d7b9a0051a62788abc780

Seems okay, ws does not have to use ref

File: fableous-fe/src/hooks/useWsConn.tsx [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    };
+    const teardownPing = () => {
+      clearInterval(interval);
+    };
+    wsConn.addEventListener("open", setupPing);
+    wsConn.addEventListener("error", teardownPing);
+    wsConn.addEventListener("close", teardownPing);
+    return () => {
+      wsConn.removeEventListener("open", setupPing);
+      wsConn.addEventListener("error", teardownPing);
```
Comment (daystram) [inline fableous-fe/src/hooks/useWsConn.tsx:28]: removeEventListener?

Comment (nicolauscg) [reply fableous-fe/src/hooks/useWsConn.tsx]: done

**Comment by nicolauscg** [comment]
UI is still bad, will look at existing pages to make it fit better, but functionality wise should be okay

**Comment by daystram** [comment]
looks good so far

File: fableous-fe/src/components/canvas/Canvas.tsx [hunk 2]
```diff
@@ -566,19 +570,33 @@ const Canvas = forwardRef<HTMLCanvasElement, CanvasProps>(
...
+      setAudioB64Strings([]);
+      setTextShapes({});
       // eslint-disable-next-line react-hooks/exhaustive-deps
-    }, [role, wsRef]);
+    }, [pageNum]);
+
+    // temporary workaround to recalculate width when canvas appears or becomes hidden
+    useEffect(() => {
+      const canvas = canvasRef.current;
+      canvas.width = canvas.offsetWidth * SCALE;
```
Comment (daystram) [inline fableous-fe/src/components/canvas/Canvas.tsx:596]: extract this canvas resize and onmount useEffect to a separate func?

Comment (nicolauscg) [reply fableous-fe/src/components/canvas/Canvas.tsx]: done

**Comment by nicolauscg** [comment]
ready for review now @daystram , can be merged if all looks good, one note:

- As I was playing with showing and hidding canvas and cursor screen, I noticed that on controller, cursor width is 0 until my mouse moves inside of it. 
  - Your workaround using targetCanvasRef on cursor screen didn't work well because changes to property of a ref does not cause rerender
  - which is why the problem is fixed when I move my mouse pointer inside (causes rerender which causes width recalc)
  - so instead in https://github.com/deco-finter/fableous/pull/14/commits/ff04f1e1ed9c7c4b1c2cb9f8d3003574baae27e9 I used a similar apporach to Canvas, using a isShown boolean prop, I checked and all dimension of canvas and cursor screen on hub and controller are identical.

**Comment by nicolauscg** [comment]
1 more thing, I also added a notistack library on commits https://github.com/deco-finter/fableous/pull/14/commits/d2d295e63ea5f3e2988e8f0a000fe381e1334804 and https://github.com/deco-finter/fableous/pull/14/commits/d0eb1b818fc4638f690890b43fd1ad4632213df8, I think its a good way to give feedback to user on errors without having to manage isOpen states when using [Mui snackbars](https://material-ui.com/components/snackbars/#complementary-projects), look at their github for a demo.

I noticed u used <Alert /> that does not disappear on classroom list page if get req returns error, I think that could stay that way, however for scenarios like when controller inputs invalid token in a form, a error msg that dissapears in a few sec seems more fitting to me

**Comment by daystram** [comment]
> ready for review now @daystram , can be merged if all looks good, one note:
> 
> * As I was playing with showing and hidding canvas and cursor screen, I noticed that on controller, cursor width is 0 until my mouse moves inside of it.
>   
>   * Your workaround using targetCanvasRef on cursor screen didn't work well because changes to property of a ref does not cause rerender
>   * which is why the problem is fixed when I move my mouse pointer inside (causes rerender which causes width recalc)
>   * so instead in [ff04f1e](https://github.com/deco-finter/fableous/commit/ff04f1e1ed9c7c4b1c2cb9f8d3003574baae27e9) I used a similar apporach to Canvas, using a isShown boolean prop, I checked and all dimension of canvas and cursor screen on hub and controller are identical.

the isShown prop to toggle the canvas visibility appears to rerender the canvas and thus reset its dimensions. the original fix solves the problem before this flag was added. this new solution appears to be more consistent, lets stick with this

**Comment by daystram** [comment]
> 1 more thing, I also added a notistack library on commits [d2d295e](https://github.com/deco-finter/fableous/commit/d2d295e63ea5f3e2988e8f0a000fe381e1334804) and [d0eb1b8](https://github.com/deco-finter/fableous/commit/d0eb1b818fc4638f690890b43fd1ad4632213df8), I think its a good way to give feedback to user on errors without having to manage isOpen states when using [Mui snackbars](https://material-ui.com/components/snackbars/#complementary-projects), look at their github for a demo.
> 
> I noticed u used `<Alert />` that does not disappear on classroom list page if get req returns error, I think that could stay that way, however for scenarios like when controller inputs invalid token in a form, a error msg that dissapears in a few sec seems more fitting to me

aight, the notifications could be used for other submission form feedbacks

**Comment by daystram** [comment]
we could also allow disconnected controllers to rejoin the session. a slight change must be added where the WSMessageTypeControl message emitted after joining will include the current page, allowing the joining controller to follow the current page of the session

**Review by daystram — APPROVED** [review]

**Comment by nicolauscg** [comment]
points to address after review by Danny, will be done in another PR and this one merged as is, since this contains quite numerous changes, which may benefit Mike and Will. Nic might also slow down due to other assignments :)

- on controller side, if hub dc, tell it to user and show join form again
- on controller side, account for WS control msg having new page number data
- on hub side, if a controller disconnects, show snackbar

**Comment by nicolauscg** [comment]
> we could also allow disconnected controllers to rejoin the session. a slight change must be added where the WSMessageTypeControl message emitted after joining will include the current page, allowing the joining controller to follow the current page of the session

yup nice idea, we should implement this

**Pull request merged**